### PR TITLE
Cross thread adding

### PR DESCRIPTION
**Pull request by postwait**

### CODE REVIEW

File: src/eventer/eventer_epoll_impl.c [hunk 1]
```diff
@@ -248,14 +248,31 @@ static void eventer_epoll_impl_trigger(eventer_t e, int mask) {
...
   const char *cbname;
   ev_lock_state_t lockstate;
+  int cross_thread = mask & EVENTER_CROSS_THREAD_TRIGGER;
 
   mask = mask & ~(EVENTER_RESERVED);
   fd = e->fd;
-  if(e != master_fds[fd].e) return;
+  if(cross_thread) {
+    if(master_fds[fd].e != NULL) {
+      mtevL(mtev_error, "Attempting to trigger alread-registered event fd: %d cross thread.\n", fd);
```
Comment (aidancully) [inline src/eventer/eventer_epoll_impl.c]: typo: "alread-registered"

File: src/eventer/eventer_epoll_impl.c [hunk 2]
```diff
@@ -248,14 +248,31 @@ static void eventer_epoll_impl_trigger(eventer_t e, int mask) {
...
+    }
+    /* assert(master_fds[fd].e == NULL); */
+  }
   if(!pthread_equal(pthread_self(), e->thr_owner)) {
+    /* If we're triggering across threads, it can't be registered yet */
+    if(master_fds[fd].e != NULL) {
+      mtevL(mtev_error, "Attempting to trigger alread-registered event fd: %d cross thread.\n", fd);
```
Comment (aidancully) [inline src/eventer/eventer_epoll_impl.c]: typo: "alread-registered"

File: src/eventer/eventer_kqueue_impl.c [hunk 3]
```diff
@@ -297,11 +297,27 @@ static void eventer_kqueue_impl_trigger(eventer_t e, int mask) {
 
   mask = mask & ~(EVENTER_RESERVED);
   fd = e->fd;
-  if(e != master_fds[fd].e) return;
+  if(cross_thread) {
+    if(master_fds[fd].e != NULL) {
+      mtevL(mtev_error, "Attempting to trigger alread-registered event fd: %d cross thread.\n", fd);
```
Comment (aidancully) [inline src/eventer/eventer_kqueue_impl.c]: typo: "alread-registered"

File: src/eventer/eventer_kqueue_impl.c [hunk 4]
```diff
@@ -297,11 +297,27 @@ static void eventer_kqueue_impl_trigger(eventer_t e, int mask) {
...
+    }
+    /* assert(master_fds[fd].e == NULL); */
+  }
   if(!pthread_equal(pthread_self(), e->thr_owner)) {
+    /* If we're triggering across threads, it can't be registered yet */
+    if(master_fds[fd].e != NULL) {
+      mtevL(mtev_error, "Attempting to trigger alread-registered event fd: %d cross thread.\n", fd);
```
Comment (aidancully) [inline src/eventer/eventer_kqueue_impl.c]: typo: "alread-registered"

File: src/eventer/eventer_ports_impl.c [hunk 5]
```diff
@@ -223,14 +223,31 @@ eventer_ports_impl_trigger(eventer_t e, int mask) {
...
   struct timeval __now;
   int fd, newmask;
+  int cross_thread = mask & EVENTER_CROSS_THREAD_TRIGGER;
 
   mask = mask & ~(EVENTER_RESERVED);
   fd = e->fd;
-  if(e != master_fds[fd].e) return;
+  if(cross_thread) {
+    if(master_fds[fd].e != NULL) {
+      mtevL(mtev_error, "Attempting to trigger alread-registered event fd: %d cross thread.\n", fd);
```
Comment (aidancully) [inline src/eventer/eventer_ports_impl.c]: typo: "alread-registered"

File: src/eventer/eventer_ports_impl.c [hunk 6]
```diff
@@ -223,14 +223,31 @@ eventer_ports_impl_trigger(eventer_t e, int mask) {
...
+    }
+    /* assert(master_fds[fd].e == NULL); */
+  }
   if(!pthread_equal(pthread_self(), e->thr_owner)) {
+    /* If we're triggering across threads, it can't be registered yet */
+    if(master_fds[fd].e != NULL) {
+      mtevL(mtev_error, "Attempting to trigger alread-registered event fd: %d cross thread.\n", fd);
```
Comment (aidancully) [inline src/eventer/eventer_ports_impl.c]: typo: "alread-registered"

File: src/eventer/eventer_ports_impl.c [hunk 7]
```diff
@@ -223,14 +223,31 @@ eventer_ports_impl_trigger(eventer_t e, int mask) {
...
+    if(master_fds[fd].e != NULL) {
+      mtevL(eventer_deb, "Attempting to trigger already-registered event fd: %d cross thread.\n", fd);
+    }
+    /* assert(master_fds[fd].e == NULL); */
+  }
   if(!pthread_equal(pthread_self(), e->thr_owner)) {
+    /* If we're triggering across threads, it can't be registered yet */
+    if(master_fds[fd].e != NULL) {
+      mtevL(eventer_deb, "Attempting to trigger already-registered event fd: %d cross thread.\n", fd);
+    }
```
Comment (aidancully) [inline src/eventer/eventer_ports_impl.c]: should there be a `return` here?

File: src/eventer/eventer_ports_impl.c [hunk 8]
```diff
@@ -223,14 +223,31 @@ eventer_ports_impl_trigger(eventer_t e, int mask) {
...
+    /* assert(master_fds[fd].e == NULL); */
+
     eventer_cross_thread_trigger(e,mask);
     return;
   }
+  if(master_fds[fd].e == NULL) {
+    master_fds[fd].e = e;
+    e->mask = 0;
```
Comment (aidancully) [inline src/eventer/eventer_ports_impl.c]: why is `e->mask` cleared here? should it be cleared before `master_fds` gets updated?

Comment (aidancully) [inline src/eventer/eventer_ports_impl.c]: should the `master_fd(fd)` lock be held while `master_fds[fd]` gets updated?

Comment (postwait) [inline src/eventer/eventer_ports_impl.c]: we know we're in the thread that owns that fd, we should be safe without the lock.

File: src/eventer/eventer_ports_impl.c (hunk 7 above)
Comment (postwait) [inline src/eventer/eventer_ports_impl.c]: not if we want it to behave like it used to... that should become an assert at some point later, but now, we should carry on and risk it.

Comment (aidancully) [inline src/eventer/eventer_ports_impl.c]: OK, I see now - if we called `eventer_add` before `eventer_trigger`, the `eventer_t` will likely be in `master_fds` during `eventer_cross_thread_process`.

**Comment by aidancully** [comment]
+1

**Pull request merged**